### PR TITLE
Don't create `DeviceArrays` for `onp.ndarray` indices

### LINKED ISSUE #2878: _index_to_gather produces DeviceArrays for int / onp.ndarray inputs
The `_index_to_gather` function and some subfunctions, like `_normalize_index`, perform operations with `lax` functions even when dealing with integer or onp.ndarray inputs. That's unfortunate because it means, for example, doing scalar integer operations on the default accelerator device! We should probably revise.

### PR DESCRIPTION
**Pull request by lgeiger**
This PR is a follow up to #10393 and prevents the creation of `DeviceArrays` when indexing with `onp.ndarray`. This prevents index computations to happen on accelerators if the index is an `onp.ndarray`.

```python
import numpy as np
from jax import make_jaxpr, numpy as jnp

x = jnp.arange(10)
idx = np.array([1, 2, 3])
print(make_jaxpr(lambda: x[idx])())
```
would previously generate:
```
{ lambda a:i32[3] b:i32[10]; . let
    c:bool[3] = lt a 0
    d:i32[3] = add a 10
    e:i32[3] = select_n c a d
    f:i32[3,1] = broadcast_in_dim[broadcast_dimensions=(0,) shape=(3, 1)] e
    g:i32[3] = gather[
      dimension_numbers=GatherDimensionNumbers(offset_dims=(), collapsed_slice_dims=(0,), start_index_map=(0,))
      fill_value=None
      indices_are_sorted=False
      mode=GatherScatterMode.PROMISE_IN_BOUNDS
      slice_sizes=(1,)
      unique_indices=False
    ] b f
  in (g,) }
```
With this PR index normalisations happens in native Numpy so the Jaxpr becomes:
```
{ lambda a:i32[10] b:i32[3,1]; . let
    c:i32[3] = gather[
      dimension_numbers=GatherDimensionNumbers(offset_dims=(), collapsed_slice_dims=(0,), start_index_map=(0,))
      fill_value=None
      indices_are_sorted=False
      mode=GatherScatterMode.PROMISE_IN_BOUNDS
      slice_sizes=(1,)
      unique_indices=False
    ] a b
  in (c,) }
```

If I haven't missed anything this should fix #2878.

### CODE REVIEW

**Comment by jakevdp** [comment]
I don't think we should do this, because this is a recipe for inadvertently generating very large compile-time constants that can slow down tracing a lot. Similar behavior that we'd like to get rid of can be seen here: #10596

**Comment by YouJiacheng** [comment]
@jakevdp Hmmm, this PR can slow down tracing since it normalize idx during tracing.
But IIUC it won't generate larger compile-time constant than before:
only from un-normalized compile-time constant `np.ndarray` to normalized compile-time constant `np.ndarray` with same size.

**Comment by YouJiacheng** [comment]
Nice explanation, I think it is good to keep this convention(avoid trace-time op on np array)!!👍😊
